src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ingroovy/GroovyStepExecution.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    public boolean start() throws Exception {
+        CpsStepContext cps = getContext();
+        CpsThread t = CpsThread.current();
+
+        // TODO: make sure this class is actually CPS transformed
+        // it's possible that the plugin dev has messed up and put the subtype in src/main/groovy,
+        // ... in which case the class will resolve but will not work as expected.
+
+        Closure body = InvokerHelper.getMethodPointer(this, "call");
+        if (getStep().getDescriptor().takesImplicitBlockArgument()) {
+            if (body.getMaximumNumberOfParameters()==0)
+                throw new IllegalArgumentException(getClass().getName()+" claims to take the body block, but its call method takes no argument.");
+            body = body.curry(cps.newBodyInvoker().getBody());
+        }
+
+        execution = cps.newBodyInvoker(t.getGroup().export(body))
+//                .withStartAction(/*... maybe a marker for the future?*/)
+                .withCallback(BodyExecutionCallback.wrap(cps))
+                .withoutBlockNode()
+                .start();
+
+        return false;
+    }
+
+    @Override
+    public void stop(@Nonnull Throwable cause) throws Exception {
+        if (execution!=null)
+            execution.cancel(cause);
+    }
+
+    /**
+     * Attempt to resolve a method call as a step invocation like Pipeline Script.
+     *
+     * See http://groovy-lang.org/metaprogramming.html#_methodmissing
+     */
+    public Object methodMissing(String name, Object args) throws IOException {
+        GroovyObject d = getDelegate();
+        if (d!=null)
+            return d.invokeMethod(name,args);
+        else
+            throw new MissingMethodException(name,getClass(),unpack(args));
+    }
+
+    private Object[] unpack(Object args) {
+        if (args instanceof Object[])
+            return (Object[]) args;
+        if (args==null)
+            return new Object[0];
+        else
+            return new Object[]{args};
+    }
+
+    /**
+     * Attempt to resolve a property access as a step invocation like Pipeline Script.
+     *
+     * See http://groovy-lang.org/metaprogramming.html#_propertymissing
+     */
+    public Object propertyMissing(String name) throws IOException {
+        GroovyObject d = getDelegate();
+        if (d!=null)
+            return d.getProperty(name);
+        else
+            throw new MissingPropertyException(name,getClass());
+    }
+
+    /**
+     * Delegate that implements the method/property resolution of Pipeline Script.
+     *
+     * <p>
+     * This needs to happen inside CPS VM to work correctly, hence no caching.
+     *
+     * @return null
+     *      if this method is invoked (incorrectly) outside the CPS VM thread.
+     */
+    protected @Nullable GroovyObject getDelegate() throws IOException {
+        if (CpsThread.current()==null)
+            return null;    // invocation outside CPS VM thread
+        return new CpsScript() {
```

Comment:
So this is definitely what's causing the classloader problems I mentioned. Since this `CpsScript` instance is being instantiated outside of the `GroovyShell`, it's not getting the right `ClassLoader`, and anything that's trying to use `script.getClass().getClassLoader()`, like [this example in `DockerDSL`](https://github.com/jenkinsci/docker-workflow-plugin/blob/master/src/main/java/org/jenkinsci/plugins/docker/workflow/DockerDSL.java#L47), will not work. 

I haven't been able to figure out how to solve this - there's no way I can find to get at the existing `CpsScript` instance at runtime to just pass that back onward, so...yeah. Confusion. Problems. And a :bug:. =)